File: DataStructure/LinkedList/LinkedList.cpp
#include <iostream>
using namespace std;

//Linked List Implemention in C++

class LinkedList {
    private:
        // This module defines a simple linked list sturcture with basic operations.
        class Node{
            public:
                int data;
                Node* next;

                //Declare a Node consisting of data and pointer to the next node
                Node(int data) : data(data), next(nullptr) {}
        };

        Node* head;
        int size;

    public:
        LinkedList() : head(nullptr), size(0) {}

        // Method to insert a new node at the end of the list
        void insert(int data) {
            Node* newNode = new Node(data); // Create a new node with the given data
            if(this->head == nullptr) { this->head = newNode; } // if head is null, set the new node as the head
            else {
                Node* current = head;// Start from the head
                while(current->next != nullptr) {  current = current->next; } // Traverse to the last node
                current->next = newNode; // Link the new node at th and
            }
            this->size++;
        }

        // Method to delete a node with a specific value
        void remove(int data) {
            if(this->head == nullptr) return;// Make sure head is empty
            if(this->head->data == data) {   // Make sure first Node equal to data
                Node* temp = head;
                this->head = this->head->next;
                delete temp;
                this->size--;
            }
            else {
                Node* current = this->head;//
                while(current->next != nullptr && current->next->data != data) { current = current->next; }
                if(current->next != nullptr) {
                    Node* temp = current->next;
                    current->next = current->next->next;
                    delete temp;
                    this->size--;
                }
            
            }
        }

        // Method to find a node with a specific value
        bool find(int data) {
            Node* current = this->head;
            while(current != nullptr) {
                if(current->data == data) { return true; }
                current = current->next;
            }
            return false;
        }

        // Method to size of the LinkedList
        int list_size() {return this->size; }

        // Method to print the LinkedList
        void print() {
            Node* current = this->head;
            while(current != nullptr) {
                cout << current->data << "->";
                current = current->next;
            }
            cout << "null" << endl;
        }
};

int main() {
    LinkedList ll;
    ll.insert(1);
    ll.insert(2);
    ll.insert(3);
    ll.insert(4);
    ll.insert(5);
    ll.print(); // 1->2->3->->null
    cout << ll.list_size() << endl; // 3
    cout << (ll.find(2) ? "found" : "not found") << endl; // found
    ll.remove(2);
    ll.print(); // 1->3->null
    cout << ll.list_size() << endl; // 2
    return 0;
}

File: DataStructure/LinkedList/LinkedList.java
//Linked List Implementation in Java

// LinkedList class to manage the linked list operations
public class LinkedList {

    //This module defines a simple linked list structure with basic operations.
    private static class Node {
        private final int data;
        private Node next;

    // Declare a Node consisting of data and a pointer to the next node
    public Node(int data) {
        this.data = data;
        this.next = null; // Initialize next as null
    }
}
    private Node head; // Head of the list
    private int size; // Size of the list

    // Constructor to initialize the linked list
    public LinkedList() {
        this.head = null; // Start with an empty list
        this.size = 0; // Size is zero initially
    }

    // Method to insert a new node at the end of the list
    public void insert(int data) {
        Node newNode = new Node(data);
        if (this.head == null) { this.head = newNode; }
        else { Node current = this.head;
            while (current.next != null) { current = current.next; }
            current.next = newNode;
        }
        this.size++; // Increment the size of the list
    }

    // Delete a node with a specific value
    public void delete(int data) {
        if (this.head.data == data) { // If the head node is to be deleted
            this.head = this.head.next; // Move head to the next node
            this.size--; // Decrement size
        } else{
            Node current = this.head;
            while (current.next != null && current.next.data != data) {
                current = current.next; // Traverse the list
            }
            if (current.next != null) { // If the node to delete was found
                current.next = current.next.next; // Bypass the node
                this.size--; // Decrement size
            }
        }
    }

    //Find a node with a specific value
    public boolean find(int data) {
        Node current = this.head;
        while(current != null) {// Traverse the list
            if(current.data == data) { return true; }
            current = current.next; // Move to the next node
        }
        return false; // Return false if not found
    }

    // Get the size of the LinkedList
    public int size() { return this.size; }

    //Method to print the LinkedList
    public void print() {
        String lstr = "";
        Node current = this.head;

        while(current != null) {
            lstr += String.valueOf(current.data);
            lstr += " -> ";
            current = current.next;
        }
        lstr += "Null";
        System.out.println(lstr);
    }

    //Example usage
    public static void main(String[] args) {
        LinkedList ll = new LinkedList();
        ll.insert(1);
        ll.insert(3);
        ll.insert(5);
        ll.insert(7);
        ll.insert(9);
        ll.print();// Output: 1 -> 3 -> 5 -> 7 -> 9 -> null
        System.out.println(ll.size());// Output: 4
        ll.delete(5);
        ll.print();// Output: 1 -> 3 -> 7 -> 9 -> null
    }
}
